### PR TITLE
[Merged by Bors] - feat(CategoryTheory): the internal hom with the monoidal unit is the identity

### PR DESCRIPTION
**Pull request by emilyriehl**
In a closed monoidal category, the internal hom defined by mapping out of the monoidal unit is naturally isomorphic to the identity. This specializes to an analogous result about exponentiating with the terminal object in a cartesian closed category.

Co-authored-by: [Dagur Asgeirsson](https://github.com/dagurtomas) 

---

UPDATE: The questions raised here have been answered in the comments below. This is now ready for review.

There are two things I don't understand about this code.

(1) Why must `MonoidalClosed.unitNatIso` be marked uncomputable?
(2) Which of `expTerminalNatIso` or `expTerminalNatIso'` is better?

For (2), I'd assumed it would be more useful not to have to supply a proof that the terminal object is exponentiable. But when I gave an explicit proof in (1) I wasn't able to extract the component used to prove `expTerminalIsoSelf`, which existed already in the form that it's currently stated in.

-->

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by dagurtomas** [comment]
For (1), it's because it uses `conjugateIsoEquiv`, which is noncomputable because it uses `asIso` which produces an isomorphism (containing the data of an inverse) just by knowing that a morphism _has_ an inverse.

For (2), I would supply `MonoidalClosed.unitNatIso` with an arbitrary instance variable `[Closed (𝟙_ C)]` instead (which will be automatically inferred whenever your in a monoidal closed category anyway) instead of using the explicit `unitClosed`. Then do the same with `Exponentiable` in the cartesian one. See [this branch](https://github.com/leanprover-community/mathlib4/compare/master...dagur/closedUnit)

**Comment by dagurtomas** [comment]
Regarding computability, note that [CategoryTheory.Adjunction.natIsoEquiv](https://leanprover-community.github.io/mathlib4_docs/Mathlib/CategoryTheory/Adjunction/Unique.html#CategoryTheory.Adjunction.natIsoEquiv) is computable, so `conjuagateIsoEquiv` shouldn't need to be noncomputable. 

These two APIs should really be merged at some point, see the TODO in the file [Adjunction.Unique](https://leanprover-community.github.io/mathlib4_docs/Mathlib/CategoryTheory/Adjunction/Unique.html). I wrote that part of the uniqueness file without knowing I was duplicating API, then when I discovered it it turned out to be nontrivial to deduplicate

**Comment by emilyriehl** [comment]
Thanks @dagurtomas. I would like to merge your branch with this one. How do I do this?

**Comment by emilyriehl** [comment]
Thanks also for pointing out the issue with the way I defined natIsoEquiv. When I get a moment (sometime this week) I'll go back into mates and do that part of the TODO from the Adjunction.Unique file.

**Comment by dagurtomas** [comment]
> I would like to merge your branch with this one. How do I do this?

```
git checkout closedUnit 
git merge dagur/closedUnit
```

**Comment by b-mehta** [comment]
The idea of having `[Exponentiable (⊤_ C)]` as an argument rather than having it be inferred is, approximately, to ensure that the proof works given _any_ proof that 1 is exponentiable. If this argument weren't given directly, then the proof would be for a particular choice of the exponential of 1. Of course any choice is naturally isomorphic to any other, but it's convenient to not need to add this isomorphism where necessary. Admittedly my docstring "The typeclass argument is explicit: any instance can be used." isn't particularly clear about this, although hopefully "any instance can be used" makes a little more sense now.

**Comment by emilyriehl** [comment]
> > I would like to merge your branch with this one. How do I do this?
> 
> ```
> git checkout closedUnit 
> git merge dagur/closedUnit
> ```

This returns the error `merge: dagur/closedUnit - not something we can merge`. I'm working from the command line on my computer in case this is relevant. I tried also running `git fetch` first but that didn't help.

**Comment by dagurtomas** [comment]
Thanks!

maintainer merge

**Comment by dagurtomas** [comment]
Regarding merging, I guess my suggestion requires you to have a local copy of my branch. So for future reference, I think it would have worked after running `git fetch` and `git checkout dagur/closedUnit` or something like that

File: Mathlib/CategoryTheory/Closed/Cartesian.lean [hunk 1]
```diff
@@ -226,21 +226,16 @@ end CartesianClosed
...
 
-/-- Show that the exponential of the terminal object is isomorphic to itself, i.e. `X^1 ≅ X`.
+/-- The exponential with the terminal object is naturally isomorphic to the identity. The typeclass
+argument is explicit: any instance can be used.-/
+def expTerminalNatIso [Exponentiable (⊤_ C)] : 𝟭 C ≅ exp (⊤_ C) :=
+  MonoidalClosed.unitNatIso (C := C)
 
-The typeclass argument is explicit: any instance can be used.
+/-- A component of the above in the reversed direction, showing that the exponential with the
+terminal object is isomorphic to itself, i.e. `X^1 ≅ X`.
```
[the remark this answers is not in the document]
Comment (emilyriehl) [reply Mathlib/CategoryTheory/Closed/Cartesian.lean]: Ah, I didn't realize. Fixed now.

**Comment by dagurtomas** [comment]
maintainer merge

**Comment by kim-em** [comment]
bors merge

**Comment by b-mehta** [comment]
> Thanks for the explanation. I guess I hadn't quite figured out how to interpret typeclass arguments. I guess the idea is that even though we might specify a particular instance of `[Exponentiable (⊤_ C)]` some other term might be used somewhere else?

Yeah essentially - we can give an instance of `Exponentiable (⊤_ C)` directly, but it could also be the case that we have that the category is cartesian closed (eg because it's equivalent to another cartesian closed one), which gives a different `Exponentiable (⊤_ C)` instance. And the point of writing the proof like this is that it works regardless of which of the above are used as the construction.

> By the way, does Lean allow you to label two distinct terms of a given type as instances?

Yes. In this example, we could have two different constructions of the right adjoint, which are not equal on the nose. It's usually not a good idea to do this, but it's possible.

**Pull request closed without merging**